### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by mathewc**
See associated SDK PR: https://github.com/Azure/azure-webjobs-sdk/pull/2550. This PR relies on that one. Once the SDK change is merged, I'll update this PR to reference the new required SDK version.

These improvements result in about an **10% RPS and 9% latency gain** in a .NET Hello World scenario.

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Diagnostics/FunctionInstanceLogger.cs [hunk 1]
```diff
@@ -80,32 +81,13 @@ internal FunctionInstanceLogger(IFunctionMetadataManager metadataManager, IMetri
...
             }
-            else
+            else if (item.IsCompleted)
             {
-                // Function Started
-                if (monitor == null)
-                {
-                    if (_metadataManager.TryGetFunctionMetadata(item.LogName, out FunctionMetadata function))
-                    {
-                        monitor = new FunctionInstanceMonitor(function, _metrics, item.FunctionInstanceId);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Diagnostics/FunctionInstanceLogger.cs:99]: Removing this FunctionInstanceMonitor type completely to remove an object allocation from the invocation path

File: src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs [hunk 2]
```diff
@@ -336,12 +336,14 @@ private class FunctionActivityTracker : IDisposable
...
             private readonly IMetricsPublisher _metricsPublisher;
             private readonly ILinuxContainerActivityPublisher _linuxContainerActivityPublisher;
+            private readonly object _runningFunctionsSyncLock = new object();
 
             private ulong _totalExecutionCount = 0;
+            private int _activeFunctionCount = 0;
             private int _functionActivityFlushInterval;
             private CancellationTokenSource _etwTaskCancellationSource = new CancellationTokenSource();
             private ConcurrentQueue<FunctionMetrics> _functionMetricsQueue = new ConcurrentQueue<FunctionMetrics>();
-            private ConcurrentDictionary<string, RunningFunctionInfo> _runningFunctions = new ConcurrentDictionary<string, RunningFunctionInfo>();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs:344]: Use of this concurrent dictionary was causing a "hot lock" to show up in profiles. I've moved to a simpler locking strategy.

File: src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs [hunk 3]
```diff
@@ -428,33 +427,27 @@ public void Dispose()
...
-                var functionStage = (startedEvent.Success == false) ? ExecutionStage.Failed : ExecutionStage.Succeeded;
+                Interlocked.Decrement(ref _activeFunctionCount);
+
+                var functionStage = !startedEvent.Success ? ExecutionStage.Failed : ExecutionStage.Succeeded;
                 long executionTimeInMS = (long)startedEvent.Duration.TotalMilliseconds;
                 var monitoringEvent = new FunctionMetrics(startedEvent.FunctionMetadata.Name, functionStage, executionTimeInMS);
                 _functionMetricsQueue.Enqueue(monitoringEvent);
 
-                var key = GetDictionaryKey(startedEvent.FunctionMetadata.Name, startedEvent.InvocationId);
-                if (_runningFunctions.TryRemove(key, out RunningFunctionInfo functionInfo))
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs:449]: We no longer remove from the collection on completion, reducing lock contention. We just raise the completed event. The background timer that is raising in progress events will skip/purge any completed functions.

File: src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs [hunk 4]
```diff
@@ -428,33 +427,27 @@ public void Dispose()
...
             {
                 _totalExecutionCount++;
+                Interlocked.Increment(ref _activeFunctionCount);
 
                 var monitoringEvent = new FunctionMetrics(startedEvent.FunctionMetadata.Name, ExecutionStage.Started, 0);
                 _functionMetricsQueue.Enqueue(monitoringEvent);
 
-                var key = GetDictionaryKey(startedEvent.FunctionMetadata.Name, startedEvent.InvocationId);
-                var triggerType = startedEvent.FunctionMetadata.Trigger?.Type;
-                var value = new RunningFunctionInfo(startedEvent.FunctionMetadata.Name, startedEvent.InvocationId, startedEvent.Timestamp, startedEvent.Success, triggerType);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs:437]: Removing this type completely. It was just a copy of data we already have in the started event, so we just track those instances directly.

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs [hunk 5]
```diff
@@ -267,7 +269,7 @@ protected override async Task<object> InvokeCore(object[] parameters, FunctionIn
             int actualParameterCount = function.GetParameters().Length;
             parameters = parameters.Take(actualParameterCount).ToArray();
 
-            object result = function.Invoke(null, parameters);
+            object result = await _methodInvoker.InvokeAsync(null, parameters);
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs:272]: Replacing reflection invoke with call to compiled lambda. We're now using the same MethodInvokerFactory that the core SDK uses when invoking methods.

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs [hunk 6]
```diff
@@ -276,12 +278,6 @@ protected override async Task<object> InvokeCore(object[] parameters, FunctionIn
                 originalParameters[i] = parameters[i];
             }
 
-            // unwrap the task
-            if (result is Task)
-            {
-                result = await ((Task)result).ContinueWith(t => GetTaskResult(t), TaskContinuationOptions.ExecuteSynchronously);
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs:282]: No longer need to use reflection to unwrap the task result

File: test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/ProxyEndToEndTests.cs [hunk 7]
```diff
@@ -28,11 +28,13 @@ public class ProxyEndToEndTests : IClassFixture<ProxyEndToEndTests.TestFixture>
     {
         private readonly ScriptSettingsManager _settingsManager;
         private TestFixture _fixture;
+        private string _hostName;
 
         public ProxyEndToEndTests(TestFixture fixture)
         {
             _fixture = fixture;
             _settingsManager = ScriptSettingsManager.Instance;
+            _hostName = new HostNameProvider(SystemEnvironment.Instance).Value ?? "localhost";
```
Comment (mathewc) [inline test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/ProxyEndToEndTests.cs:37]: Often on my machine I have a WEBSITE_SITE_NAME env var set which breaks these tests.

**Review by paulbatum — CHANGES_REQUESTED** [review]
Same feedback as on the associated webjobs PR, there's a few different changes happening here (how the instance logger works, using compiled lambdas), please send as separate PRs.

**Pull request closed without merging**